swarm/storage/mru/handler.go

```diff
@@ -0,0 +1,506 @@
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Handler is the API for Mutable Resources
+// It enables creating, updating, syncing and retrieving resources and their update data
+package mru
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/ethereum/go-ethereum/swarm/log"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+)
+
+const chunkSize = 4096 // temporary until we implement FileStore in the resourcehandler
+
+type Handler struct {
+	chunkStore      *storage.NetStore
+	HashSize        int
+	resources       map[uint64]*resource
+	resourceLock    sync.RWMutex
+	storeTimeout    time.Duration
+	queryMaxPeriods uint32
+}
+
+// HandlerParams pass parameters to the Handler constructor NewHandler
+// Signer and TimestampProvider are mandatory parameters
+type HandlerParams struct {
+	QueryMaxPeriods uint32
+}
+
+// hashPool contains a pool of ready hashers
+var hashPool sync.Pool
+var minimumChunkLength int
+
+// init initializes the package and hashPool
+func init() {
+	hashPool = sync.Pool{
+		New: func() interface{} {
+			return storage.MakeHashFunc(resourceHashAlgorithm)()
+		},
+	}
+	if minimumMetadataLength < minimumUpdateDataLength {
+		minimumChunkLength = minimumMetadataLength
+	} else {
+		minimumChunkLength = minimumUpdateDataLength
+	}
+}
+
+// NewHandler creates a new Mutable Resource API
+func NewHandler(params *HandlerParams) (*Handler, error) {
+
+	rh := &Handler{
+		resources:       make(map[uint64]*resource),
+		storeTimeout:    defaultStoreTimeout,
+		queryMaxPeriods: params.QueryMaxPeriods,
+	}
+
+	for i := 0; i < hasherCount; i++ {
+		hashfunc := storage.MakeHashFunc(resourceHashAlgorithm)()
+		if rh.HashSize == 0 {
+			rh.HashSize = hashfunc.Size()
+		}
+		hashPool.Put(hashfunc)
+	}
+
+	return rh, nil
+}
+
+// SetStore sets the store backend for the Mutable Resource API
+func (h *Handler) SetStore(store *storage.NetStore) {
+	h.chunkStore = store
+}
+
+// Validate is a chunk validation method
+// If it looks like a resource update, the chunk address is checked against the ownerAddr of the update's signature
+// It implements the storage.ChunkValidator interface
+func (h *Handler) Validate(chunkAddr storage.Address, data []byte) bool {
+
+	dataLength := len(data)
+	if dataLength < minimumChunkLength {
+		return false
+	}
+
+	//metadata chunks have the first two bytes set to zero
```

Comment:
there should be no special metadata chunk for purposes of chunk validation.
metainfo chunks are content addressed and validate as is.
In order to disallow third party squatting of mru info chunks, all you need is to have a signature in the data.
This signature is validated by  the handler when you retrieve before first looking up an update for that MRU

Reply:
We already discussed this weeks ago... 
metainfo chunks are indeed content addressed and validate as is. However, it does not hurt to check if at least those two bytes are set to zero. If they are not, then I won't spend resources hashing to fully validate them.

Reply:
yes, let's not repeat this discussion, please.

Reply:
sorry this is important. By doing this you exclude certain correct content address chunks from being valid. this is unnecessary. You can do this check within mru after the assumption that it is an mru metachunk. 
This validation does not belong here.

Reply:
Within the MRU chunkstore, with the current implementation, there cannot be chunks beginning with [0,0] that are not metachunks.
I'll move it off anyway.

Reply:
`Validate()` must return true or false if it recognizes something good, so how  the handler figures out if it has to return `true` is the `mru.Handler` business. Those two bytes being zero is a necessary condition **but not a sufficient condition** to accept the chunk.

If we return `false`, the `ContentAddressValidator` has a chance to return true. The goal is to discard non MRU chunks as quickly as possible if we clearly see it can't be a metachunk.